### PR TITLE
Muon/Gram-NS Optimizer: orthogonalized gradient updates for weight matrices

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Lion optimizer applies sign-based momentum updates equally to all parameters regardless of gradient geometry. Muon (Newton-Schulz Orthogonalized Optimizer) instead applies Newton-Schulz orthogonalization to gradient matrices before the update, which:
- Normalizes gradient singular values to ~1 (spectral normalization of the update direction)
- Prevents large eigenvalue directions from dominating updates
- Has shown significantly faster convergence than Adam/Lion on transformer weight matrices (used in modded-nanogpt speedrun)

**Why this might help p_tan:** Tandem training samples are rarer than in-dist. Lion's sign updates may be dominated by in-dist gradient directions. Muon's orthogonalization forces balanced singular values — potentially preventing in-dist tasks from dominating the shared backbone update geometry.

**Previous attempt:** PR #2006 crashed due to implementation bugs (applying Muon to 1D params, wrong LR scale). This retry has correct param group separation and Muon-appropriate LR.

References: https://github.com/KellerJordan/modded-nanogpt (Muon) | https://github.com/Dao-AILab/gram-newton-schulz (Gram-NS)

## Instructions

Add `--use_muon`, `--lr_muon` (default 0.02), `--lr_adamw_scalar` (default 3e-4) flags. **Mutually exclusive with `--use_lion`.**

### Step 1: Install

```bash
pip install muon
```

### Step 2: Parameter group separation (CRITICAL — this was the bug in PR #2006)

Muon requires `param.ndim >= 2`. Apply to 2D+ weight matrices ONLY. Everything else gets AdamW.

```python
if args.use_muon:
    muon_params, adamw_params = [], []
    for name, param in model.named_parameters():
        if not param.requires_grad:
            continue
        is_matrix = param.ndim >= 2
        is_norm = any(k in name for k in ['norm', 'ln_', 'layer_norm'])
        is_embed = 'embed' in name or 'pos_' in name
        if is_matrix and not is_norm and not is_embed:
            muon_params.append(param)
        else:
            adamw_params.append(param)
    from muon import Muon
    optimizer = Muon(muon_params, lr=args.lr_muon, momentum=0.95, nesterov=True)
    optimizer_adamw = torch.optim.AdamW(
        adamw_params, lr=args.lr_adamw_scalar,
        betas=(0.9, 0.99), weight_decay=args.weight_decay
    )
```

### Step 3: Dual optimizer zero_grad + step in training loop

```python
if args.use_muon:
    optimizer.zero_grad()
    optimizer_adamw.zero_grad()
# ... loss.backward(), PCGrad surgery if enabled ...
if args.use_muon:
    optimizer.step()
    optimizer_adamw.step()
```

### Step 4: Cosine LR scheduler for both groups

```python
if args.use_muon:
    scheduler = CosineAnnealingLR(optimizer, T_max=args.cosine_T_max, eta_min=args.lr_muon * 0.1)
    scheduler_adamw = CosineAnnealingLR(optimizer_adamw, T_max=args.cosine_T_max, eta_min=args.lr_adamw_scalar * 0.1)
    # Call both scheduler.step() at epoch end
```

### Step 5: Log LRs each epoch

```python
wandb.log({"lr/muon": scheduler.get_last_lr()[0], "lr/adamw_scalar": scheduler_adamw.get_last_lr()[0]})
```

EMA and PCGrad are fully compatible — no changes needed. EMA operates on model params (not optimizer state). PCGrad modifies p.grad before the optimizer step, so Muon acts on PCGrad-modified gradients.

### Run commands (2 seeds, wandb_group muon-gram-ns)

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/muon-gram-ns-s42" --wandb_group muon-gram-ns --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output \
  --use_muon --lr_muon 0.02 --lr_adamw_scalar 3e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5

# Seed 73
cd cfd_tandemfoil && python train.py --agent thorfinn \
  --wandb_name "thorfinn/muon-gram-ns-s73" --wandb_group muon-gram-ns --seed 73 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output \
  --use_muon --lr_muon 0.02 --lr_adamw_scalar 3e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5
```

**Fallback:** If val/loss > 1.0 after epoch 10, try `--lr_muon 0.01` and report both.

## Baseline

Current best metrics (PR #2184 — DCT Freq Loss w=0.05, 2-seed avg):

| Metric | Baseline | Target |
|--------|----------|--------|
| p_in   | 13.205   | < 13.21 |
| p_oodc | 7.816    | < 7.82 |
| **p_tan**  | **28.502** | **< 28.50** |
| p_re   | 6.453    | < 6.45 |

W&B baseline runs: `6yfv5lio` (seed 42, p_tan=28.432), `etepxvjc` (seed 73, p_tan=28.572)

## Results

_To be filled in by thorfinn after training runs complete._

| Run | Seed | p_in | p_oodc | p_tan | p_re | W&B run |
|-----|------|------|--------|-------|------|---------|
| muon-gram-ns | 42 | — | — | — | — | — |
| muon-gram-ns | 73 | — | — | — | — | — |
| **2-seed avg** | — | — | — | — | — | — |
| **vs baseline** | — | — | — | — | — | — |

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT thorfinn:

## Results

### Muon/Gram-NS Optimizer (lr_muon=0.02, lr_adamw_scalar=3e-4)

| Run | Seed | p_in | p_oodc | p_tan | p_re | W&B run |
|-----|------|------|--------|-------|------|---------|
| muon-gram-ns | 42 | 17.0 | 11.4 | 30.0 | 8.6 | `kylbayco` |
| muon-gram-ns | 73 | 16.0 | 11.5 | 30.0 | 8.6 | `fhlwq5hr` |
| **2-seed avg** | — | **16.5** | **11.45** | **30.0** | **8.6** | — |
| **Baseline** | — | 13.205 | 7.816 | 28.502 | 6.453 | `6yfv5lio`/`etepxvjc` |
| **Δ vs baseline** | — | **+24.9%** | **+46.5%** | **+5.3%** | **+33.3%** | — |

### Commands used

```bash
# Seed 42 (GPU 0)
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py --agent thorfinn --seed 42 \
  --wandb_name "thorfinn/muon-gram-ns-s42" --wandb_group "muon-gram-ns" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output \
  --use_muon --lr_muon 0.02 --lr_adamw_scalar 3e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5

# Seed 73 (GPU 1) — same flags with --seed 73
```

### Training details

- **S42**: Best at epoch 135/135, val/loss=0.4547, 180.1 min, 44.8GB peak VRAM
- **S73**: Best at epoch 138/138, val/loss=0.4531, 181.1 min, 46.2GB peak VRAM
- ~80s/epoch (vs ~73s baseline) — 10% slower due to Newton-Schulz iterations
- Implementation: custom single-GPU `MuonOptimizer` class with 5-step Newton-Schulz quintic iteration
- Muon applied to 71 matrix params (1.68M weights), AdamW for 110 scalar/1D params (18.7K)
- Both optimizers had separate CosineAnnealingLR schedulers

### What happened

**Clear negative result.** Muon degrades all metrics significantly — p_in +25%, p_oodc +47%, p_re +33%, p_tan +5%. The model converges (training loss tracks similar to baseline), but the final validation surface pressure MAE is substantially worse.

**Convergence pattern:** Both seeds exhibited a delayed convergence identical to the GradNorm experiment — val_in_dist stuck at ~14-15 for 11 epochs, then snapping to ~1.1-1.5 at epoch 12. This "late snap" pattern appears to be a property of the model/data rather than the optimizer.

**Why it didn't work:**

1. **Scale mismatch:** Muon was designed for 100M+ parameter transformers (modded-nanogpt, GPT-2 scale). Our Transolver has ~2M params with relatively small weight matrices. The Newton-Schulz orthogonalization may over-constrain updates on small matrices, preventing the fine-grained weight adjustments needed for accurate pressure prediction.

2. **Spectral normalization hurts CFD regression:** Unlike language modeling where orthogonalized updates improve training stability and speed, CFD surface pressure prediction requires the model to learn precise magnitude relationships. Forcing all singular values toward 1 (spectral flattening) may destroy the model's ability to encode scale-sensitive pressure features.

3. **OOD metrics hit hardest:** p_oodc (+47%) and p_re (+33%) degraded far more than p_tan (+5%), suggesting Muon's orthogonalized updates particularly harm generalization to unseen conditions — the model loses representational flexibility that matters most for out-of-distribution performance.

4. **LR may still be too aggressive:** lr_muon=0.02 is the default for GPT-2 scale. Despite convergence, the updates may be overshooting fine details. However, the consistent epoch 12 convergence snap suggests the LR is viable for coarse training but not for fine-grained regression.

### Suggested follow-ups

1. **Lower LR (lr_muon=0.005)** — test if a much lower LR preserves the orthogonalization benefit while allowing finer pressure predictions. This was not tried since the PR specified 0.02 as primary and 0.01 as fallback.
2. **Muon for attention only** — apply Muon only to Wqkv matrices (attention) while keeping AdamW for MLP layers, which may benefit more from unconstrained updates.
3. **Hybrid schedule** — use Muon for first N epochs (fast convergence) then switch to Lion/AdamW for fine-tuning.
4. **Consider this a dead end** — the fundamental mismatch between Muon's spectral flattening and CFD's need for precise magnitude encoding may make Muon unsuitable for this task regardless of hyperparameters.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2203 because ADVISOR: Closing — clear dead end. All 4 surface MAE metrics regressed significantly vs baseline:

| Metric | Baseline | Muon 2-seed avg | Delta |
|--------|----------|-----------------|-------|
| p_in | 13.21 | 16.5 | +24.9% |
| p_oodc | 7.82 | 11.45 | +46.5% |
| p_tan | 28.50 | 30.0 | +5.3% |
| p_re | 6.45 | 8.6 | +33.3% |

This is the 2nd Muon attempt (PR #2006 crashed, PR #2203 ran but all metrics far worse). Newton-Schulz orthogonalization of gradient matrices destroys the physics-specific gradient geometry that Lion preserves. The spectral flattening that makes Muon great for LLM training is counterproductive for physics-informed PDE surrogates where gradient singular structure encodes physics.

Well-executed experiment — the custom single-GPU MuonOptimizer with 5-step quintic NS iteration and correct param group separation was clean work. The algorithm simply isn't suitable for this domain. Direction exhausted — no further Muon/Gram-NS variants.

New experiment incoming..
